3.preprocessing_features/scripts/1.sc_quality_control.py

```diff
@@ -0,0 +1,322 @@
+metadata_columns = [
+    "Image_Metadata_Plate",
+    "Image_Metadata_Well",
+    "Image_Metadata_Site",
+    "Metadata_Nuclei_Location_Center_X",
+    "Metadata_Nuclei_Location_Center_Y",
+]
+
+
+# ## Load in plate to perform QC on
+
+# In[5]:
+
+
+# List all converted parquet files
+all_files = list(data_dir.glob("*_converted.parquet"))
+
+# Sort the files based on the plate name (assuming the plate name is the prefix before the first '_')
+all_files.sort(
+    key=lambda f: int(
+        f.stem.split("_")[0][2:]
+    )  # Assuming plate names are prefixed with 'BR' and followed by numbers
+)
+
+# Initialize an empty dictionary to hold DataFrames for each plate
+plates_dict = {}
+
+# Process each file in the sorted list
+for file_path in all_files:
+    start_time = time.time()  # Start timer for loading
+
+    # Extract plate name for logging and as dictionary key
+    plate_name = file_path.stem.split("_")[0]
+
+    # Load and compute the DataFrame
+    plate_df = dd.read_parquet(file_path, engine="pyarrow").compute()
+
+    end_time = time.time()  # End timer for loading
+    print(
+        f"Loaded plate: {plate_name}, Shape: {plate_df.shape}, Time taken: {end_time - start_time:.2f} seconds"
+    )
+
+    # Store the DataFrame in the dictionary with the plate name as the key
+    plates_dict[plate_name] = plate_df
+
+# Check if no files were found and loaded
+if not plates_dict:
+    print("No parquet files found.")
+
+
+# In[6]:
+
+
+# Loop through each plate dataframe in plates_dict
+for plate_name, plate_df in plates_dict.items():
+    # Print the plate name being processed
+    print(f"Processing plate: {plate_name}")
+
+    # Find large nuclei outliers for the current plate
+    large_nuclei_high_int_outliers = find_outliers(
+        df=plate_df,
+        metadata_columns=metadata_columns,
+        feature_thresholds={
+            "Nuclei_AreaShape_Area": 2,
+            "Nuclei_Intensity_IntegratedIntensity_CorrDNA": 3,
+        },
+    )
+
+    # Save large nuclei scatterplot
+    plot_large_nuclei_outliers(
+        plate_df=plate_df,
+        outliers_df=large_nuclei_high_int_outliers,
+        plate_name=plate_name,
+        qc_fig_dir=qc_fig_dir,
+    )
+
+    # Find low nuclei solidity outliers for the current plate
+    solidity_nuclei_outliers = find_outliers(
+        df=plate_df,
+        metadata_columns=metadata_columns,
+        feature_thresholds={
+            "Nuclei_AreaShape_Solidity": -2,
+        },
+    )
+
+    # Save low nuclei solidity histogram
+    plot_nuclei_solidity_histogram(
+        plate_df=plate_df,
+        outliers_df=solidity_nuclei_outliers,
+        plate_name=plate_name,
+        qc_fig_dir=qc_fig_dir,
+    )
+
+    # Assuming nuclei_outliers_df and cells_outliers_df have the same index
+    outlier_indices = pd.concat(
+        [large_nuclei_high_int_outliers, solidity_nuclei_outliers]
+    ).index
+
+    # Remove rows with outlier indices from plate_4_df
```

Comment:
What is the reason for this specific plate?

Reply:
OOPS! Great catch. That is an artifact from another project, so I will remove it. Thanks!